ds3231/ds3231.go

```diff
@@ -0,0 +1,164 @@
+// Package ds3231 provides a driver for the DS3231 RTC
+//
+// Datasheet:
+// https://datasheets.maximintegrated.com/en/ds/DS3231.pdf
+package ds3231
+
+import (
+	"machine"
+	"time"
+)
+
+type Mode uint8
+
+// Device wraps an I2C connection to a DS3231 device.
+type Device struct {
+	bus machine.I2C
+}
+
+// New creates a new DS3231 connection. The I2C bus must already be
+// configured.
+//
+// This function only creates the Device object, it does not touch the device.
+func New(bus machine.I2C) Device {
+	return Device{
+		bus: bus,
+	}
+}
+
+// Configure sets up the device for communication
+func (d *Device) Configure() bool {
+	return true
+}
+
+// IsDateTimeValid return true/false is the datetime in the device is valid
+func (d *Device) IsDateTimeValid() bool {
+	data := []byte{0}
+	err := d.bus.ReadRegister(Address, REG_STATUS, data)
+	if err != nil {
+		return false
+	}
+	return (data[0] & (1 << OSF)) == 0x00
+}
+
+// IsRunning returns if the oscillator is running
+func (d *Device) IsRunning() bool {
+	data := []uint8{0}
+	err := d.bus.ReadRegister(Address, REG_CONTROL, data)
+	if err != nil {
+		return false
+	}
+	return (data[0] & (1 << EOSC)) == 0x00
+}
+
+// SetRunning starts the internal oscillator
+func (d *Device) SetRunning(isRunning bool) error {
+	data := []uint8{0}
+	err := d.bus.ReadRegister(Address, REG_CONTROL, data)
+	if err != nil {
+		return err
+	}
+	if isRunning {
+		data[0] &^= uint8(1 << EOSC)
+	} else {
+		data[0] |= 1 << EOSC
+	}
+	err = d.bus.WriteRegister(Address, REG_CONTROL, data)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// SetDateTime sets the date and time in the DS3231
+func (d *Device) SetDateTime(dt time.Time) error {
+	data := []byte{0}
+	err := d.bus.ReadRegister(Address, REG_STATUS, data)
+	if err != nil {
+		return err
+	}
+	data[0] &^= 1 << OSF
+	err = d.bus.WriteRegister(Address, REG_STATUS, data)
+	if err != nil {
+		return err
+	}
+
+	data = make([]uint8, 7)
+	data[0] = uint8ToBCD(uint8(dt.Second()))
+	data[1] = uint8ToBCD(uint8(dt.Minute()))
+	data[2] = uint8ToBCD(uint8(dt.Hour()))
+
+	year := uint8(dt.Year() - 2000)
+	centuryFlag := uint8(0)
+	if year >= 100 {
+		year -= 100
+		centuryFlag = 1 << 7
+	}
+
+	data[3] = uint8ToBCD(uint8(dt.Weekday()))
+	data[4] = uint8ToBCD(uint8(dt.Day()))
+	data[5] = uint8ToBCD(uint8(dt.Month()) | centuryFlag)
+	data[6] = uint8ToBCD(year)
+
+	err = d.bus.WriteRegister(Address, REG_TIMEDATE, data)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetDateTime returns the date and time
+func (d *Device) GetDateTime() (dt time.Time, err error) {
```

Comment:
The convention in Go is to not provide `Get*` getters, but instead simply the value name.
However, this is not just a getter. In other drivers I've used `Read*` for that instead, to make it clear that it actually does I/O and thus is not a cheap operation. So I would propose `ReadDateTime` instead (or `ReadTime`, see the other comment).